### PR TITLE
Read letters from queue

### PR DESCRIPTION
**Pull request by rkondratowicz**
This pull request deals with reading messages from the queue and mapping them to letter only.

Read message from queue, if it's a valid Letter complete the message, otherwise send it to deadletter.

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/slc/services/SendLetterJob.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Scheduled(fixedDelay = 30_000)
+    public void run() {
+        processor.handle(msg -> {
+            try {
+                Letter letter = letterMapper.from(msg);
+                logger.info("Processing letter " + letter);
+                // TODO: generate PDF
+                // TODO: send PDF to Xerox
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/slc/services/SendLetterJob.java:37]: This is the case for using interfaces; where we have a client with preferred API and an unknown implementation that could be changed later.

TODOs are perfectly fine, too. :+1:

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/IReceiverProvider.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.hmcts.reform.slc.services.servicebus;
+
+import com.microsoft.azure.servicebus.IMessageReceiver;
+
+import java.util.function.Supplier;
+
+@FunctionalInterface
+public interface IReceiverProvider extends Supplier<IMessageReceiver> {
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/IReceiverProvider.java]: Can't see the value of it yet... Why to implement `IReceiverProvider` when `Supplier<IMessageReceiver>` provides the same API?

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(SendLetterJob.class);
+
+    private final IReceiverProvider receiverProvider;
+
+    public MessageProcessor(IReceiverProvider receiverProvider) {
+        this.receiverProvider = receiverProvider;
+    }
+
+    /**
+     * Reads message from a queue and send is as an argument to passed function.
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java]: Not sure if that's any better...
```
* Reads message from the queue and calls given action passing the message as an argument.
```

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        } else {
+            logger.trace("No messages to process");
+        }
+
+        try {
+            messageReceiver.close();
+        } catch (ServiceBusException e) {
+            logger.error("Error closing connection");
+        }
+    }
```
Comment (KarolJastrzebski) [inline src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java:70]: Hmm... When squinting eyes, a pattern emerges:

```java
try {
    messageReceiver.method(args);
} catch (InterruptedExecution | ServiceBusExecution e) {
    logger.error(message, e);
}
```

Should we leave it as is, or want to be uber-squeaky clean and refactor it?

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/IReceiverProvider.java (hunk 2 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/IReceiverProvider.java]: so that I can declare a `Component`

Comment (KarolJastrzebski) [reply src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/IReceiverProvider.java]: Would it be much of a difference if class implemented Supplier, like:

```java
@Component
public class ReceiverProvider implements Supplier<IMessageReceiver> {
```
?

Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/IReceiverProvider.java]: looks like it works. removing interface.

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.slc.services.SendLetterJob;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+@Component
+public class MessageProcessor {
+    private static final Logger logger = LoggerFactory.getLogger(SendLetterJob.class);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java]: incorrect logger class?

File: src/main/java/uk/gov/hmcts/reform/slc/model/Letter.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Override
+    public String toString() {
+        return "Letter{"
+            + "template='" + template + '\''
+            + ", values=" + values
+            + ", type='" + type + '\''
+            + '}';
+    }
+}
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/slc/model/Letter.java:31]: makes me wonder whether it was better to keep everything in one repository so these common things are in one place

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            logger.error("Unable to connect to Service Bus", e);
+            return;
+        }
+
+        IMessage message = null;
+        try {
+            message = messageReceiver.receive();
+        } catch (InterruptedException | ServiceBusException e) {
+            logger.error("Unable to read message from queue", e);
+        }
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java:43]: should we `return` here too? we are catching exception so not like just a normal return of null as a message, which later gets reported as a trace

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/InvalidMessageException.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+public class InvalidMessageException extends RuntimeException {
+
+    public InvalidMessageException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public InvalidMessageException(String message) {
+        super(message);
+    }
+}
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/InvalidMessageException.java]: can we move exceptions to `exception` package please?

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    public Letter from(IMessage msg) {
+        try {
+            Letter letter = objectMapper.readValue(msg.getBody(), Letter.class);
+            if (letter != null) {
+                return letter;
+            } else {
+                throw new InvalidMessageException("Empty message " + msg.getMessageId());
+            }
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java:26]: although it is perfectly fine to have like that, could get rid of extra else block by just comparing whether letter is null

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+     * Reads message from a queue and passes is as an argument to action.
+     */
+    public void handle(Function<IMessage, MessageHandlingResult> action) {
+
+        IMessageReceiver messageReceiver;
+
+        try {
+            messageReceiver = receiverProvider.get();
+        } catch (ConnectionException e) {
+            logger.error("Unable to connect to Service Bus", e);
```
Comment (doncem) [inline src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java:34]: It's kind of wrapping same error report as provider creates one. could follow the pattern and catch same exceptions here and get rid off _wrapper_ exception

**Review by doncem — CHANGES_REQUESTED** [review]
not picking or anything, but sometimes spacing seems to be off. comparing to like generic style.

just a minor requirement from me: keep exceptions in separate package. could review packaging later but exceptions are obvious.
also if you find the time - consider karol's comment: would be nice to reduce that repetitive try/catch block

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java (hunk 10 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java:34]: unfortunately `InterruptedException` and `ServiceBusException` are checked exceptions, so I can't

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java (hunk 7 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java:43]: no, we still need to close the connection.

File: src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java (hunk 9 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/steps/maptoletter/LetterMapper.java:26]: I avoid using `if` without `else` when only one of 2 things is supposed to happen

File: src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java (hunk 4 above)
Comment (rkondratowicz) [reply src/main/java/uk/gov/hmcts/reform/slc/services/servicebus/MessageProcessor.java:70]: I'll consider adding some syntactic sugar / DSL on a future pull request.

**Review by doncem — COMMENTED** [review]
what i meant was `uk.gov.hmcts.reform.slc.exception` package 😉

**Review by doncem — APPROVED** [review]

**Pull request merged**